src/app/core/auth-validation.service.ts

```diff
@@ -0,0 +1,123 @@
+import { Injectable } from '@angular/core';
+import { FormControl, Validators } from '@angular/forms';
+import { BehaviorSubject, Observable} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthValidationService {
+
+  public isError = new BehaviorSubject<boolean>(false);
+
+  email: FormControl = new FormControl('', [
+    Validators.required,
+    Validators.email
+  ]);
+  username: FormControl = new FormControl('', [
+    Validators.required,
+    Validators.minLength(2),
+    Validators.maxLength(30)
+  ]);
+  password: FormControl = new FormControl('', [
+    Validators.minLength(8),
+    //one number, one lower, one upper, one special, no spaces
+    Validators.pattern('^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&()+=])(?=\\S+$).*$'),
+    Validators.required
+  ]);
+  text: FormControl = new FormControl('');
+  constructor() { }
+
+  /**
+   * returns a form control object for a specific type
+   * of input field
+   *
+   * @param type the type of form control required i.e.
+   * userName, password, email, or text(defualt no validation)
+   * @returns Form control object for specific type of input field
+   */
+  public getInputFormControl(type: 'email' | 'username' | 'password' | 'text') {
+    switch(type){
+      case 'username':
+        return this.username;
+      case 'email':
+        return this.email;
+      case 'password':
+        return this.password;
+      case 'text':
+        return this.text;
+    }
+  }
+
+  /**
+   * takes a form control object, and returns a (string) error message for
```

Comment:
Are you not allowed to annotate return types instead of doing it here? i.e. public getInputErrorMessage(control: FormControl): String { ... }